core/lib/merkle_tree/src/pruning.rs

```diff
@@ -24,6 +24,28 @@ impl MerkleTreePrunerHandle {
     }
 }
 
+/// objects implementing this trait can be passed to pruner main loop, they act as a source of info about up to which version the tree should be pruned
+pub trait RetainedVersionSource {
```

Comment:
This trait and its implementations look overly complicated. What I have in mind:

- Rework the pruning logic in the Merkle tree crate a bit more, so that it receives *commands* to prune past tree versions (probably via `MerkleTreePrunerHandle`). Commands are executed asynchronously, and we're OK if the pruner is shut down in the middle of a command; we'd just send a new command with a greater pruned version after a restart. To summarize, the difference with your approach is that `MerkleTreePruner` wouldn't drive pruning; external logic would.
- Use this logic straightforwardly in the metadata calculator: obtain a pruner handle, then send it to a Tokio task which would periodically query Postgres for new data to be pruned and send corresponding commands via the handle. Spawn a pruner itself in a blocking Tokio task. (So, 2 new Tokio tasks in total.) Manage both these tasks like other component tasks (i.e., make them cancel-aware etc.; `MerkleTreePruner::run()` should already be cancel-aware provided the handle is timely dropped). It could make sense to do this not in the `metadata_calculator` code, but rather in a higher-level glue code; compare with how `LazyAsyncTreeReader` can be obtained from `MetadataCalculator` via `tree_reader()` and then awaited for / used in other tasks.

Reply:
corrected